### PR TITLE
SafeSend Dapp implementation

### PR DESCRIPTION
**Pull request by austinbourdier**
### Feature

- Implement SafeSend Dapp [#299](https://github.com/MyEtherWallet/MyEtherWallet/pull/298)

Implements the Coral Protocol working version of the SafeSend Dapp

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -23,10 +23,10 @@
 - Implement Buying Subdomains dapps. [#122](https://github.com/MyEtherWallet/MyEtherWallet/pull/122)
 - Add support for Ledger live paths. [#216](https://github.com/MyEtherWallet/MyEtherWallet/pull/216)
 - Implement Swap [#127](https://github.com/MyEtherWallet/MyEtherWallet/pull/127)
+- Implement SafeSend Dapp [#299](https://github.com/MyEtherWallet/MyEtherWallet/pull/299)
 
 ### Devop
 
-- Prevent excessive polling of eth_getTransactionReceipt [#266](https://github.com/MyEtherWallet/MyEtherWallet/pull/266)
```
Comment (gamalielhere) [inline CHANGELOG.md]: Please don't remove logs

File: package.json [hunk 2]
```diff
@@ -12,17 +12,14 @@
     "dev": "vue-cli-service serve",
     "lint:code": "vue-cli-service lint",
     "lint:md": "remark . -o -f",
-    "lint:styles": "prettier --write 'src/**/*.{scss,css}' && stylelint 'src/**/*.{scss,css}' --fix",
+    "lint:styles": "prettier --write 'src/**/*.scss' && stylelint 'src/**/*.scss' --fix",
```
Comment (gamalielhere) [inline package.json]: Undo this change, we have some css files we want linted.

File: package.json [hunk 3]
```diff
@@ -12,17 +12,14 @@
...
     "lint:tests": "eslint --fix tests/",
     "start": "npm run update && npm run dev",
     "test:unit": "vue-cli-service test:unit",
     "update:packages": "node package-test.js",
     "update:lists": "node fetchEthListData.js",
     "update": "npm run update:packages && npm run update:lists"
   },
-  "gitHooks": {
```
Comment (gamalielhere) [inline package.json]: Why are you removing the git hooks and precomits?

File: src/dapps/SecureTransaction/InterfaceContainer.scss [hunk 4]
```diff
@@ -0,0 +1,35 @@
+@import '~@/scss/GlobalVariables';
+
+%interface-container {
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/InterfaceContainer.scss]: This probably shouldn't be copied here, or rename it to match your file

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 5]
```diff
@@ -0,0 +1,461 @@
+<template>
+  <div class="send-currency-container-safe-send">
+    <interface-container-title :title="$t('common.sendSafeSendTx')" />
+    <div class="send-form">
+      <div class="advanced-content safe-send-container">
+        <div class="input-container">
+          <div style="width:55%;font-size:16px">
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: Please remove inline styles

File: src/networks/nodes/index.js [hunk 6]
```diff
@@ -11,7 +11,6 @@ import etcethereumcommonwealthparity from './etc-ethereumcommonwealth-parity';
 import etcgastracker from './etc-gastracker';
 import ethethscan from './eth-ethscan';
 import ethinfura from './eth-infura';
-import ethinfuraws from './eth-infura-ws';
```
Comment (gamalielhere) [inline src/networks/nodes/index.js]: Is there any reason these were removed?

File: src/wallets/web3-provider/methods/eth_getTransactionReceipt.js [hunk 7]
```diff
@@ -1,23 +0,0 @@
-import { toPayload } from './jsonrpc';
```
Comment (gamalielhere) [inline src/wallets/web3-provider/methods/eth_getTransactionReceipt.js]: Why was this removed?

File: src/wallets/web3-provider/providers/ws-provider.js [hunk 8]
```diff
@@ -13,14 +13,11 @@ import {
...
-    const requestManager = new Web3RequestManager(
-      new Web3WSProvider(host, options)
-    );
     delete this.wsProvider['send'];
     this.wsProvider.send = (payload, callback) => {
+      const _this = this.wsProvider;
       if (_this.connection.readyState === _this.connection.CONNECTING) {
-        setTimeout(() => {
+        setTimeout(function() {
           this.wsProvider.send(payload, callback);
```
Comment (gamalielhere) [inline src/wallets/web3-provider/providers/ws-provider.js]: Use the initialized `_this` constant.

**Comment by austinbourdier** [comment]
hey @gamalielhere - I updated the pull request to pass all builds and incorporated your feedback!

**Comment by gamalielhere** [comment]
@austinbourdier can you sync it to the most recent version of our develop branch?

**Comment by austinbourdier** [comment]
@gamalielhere - I merged the parent develop into my branch. That should be reflected in this PR. Am I missing something?

**Comment by gamalielhere** [comment]
@austinbourdier it looks like you removed the nodes that we've added in the past, and there's a couple of changes that shouldn't be part of this PR since it's not relevant to your dapp like the changes in our `SendCurrencyContainer`.

File: src/partners/coral/index.js [hunk 9]
```diff
@@ -0,0 +1,619 @@
+/* eslint-disable */
```
Comment (kvhnuke) [inline src/partners/coral/index.js]: small change, can you move this file to ` src/dapps/SecureTransaction/` we like to keep all related files in one location.
also remove `/* eslint-disable */`

**Comment by austinbourdier** [comment]
@kvhnuke updated!

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+      signedTx: '',
+      resolvedAddress: ''
+    };
+  },
+  computed: {
+    ...mapGetters({
+      account: 'account',
+      gasPrice: 'gasPrice',
+      web3: 'web3',
+      wallet: 'wallet',
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: Actually, delete this. You can get the address from `web3.eth.getCoinbase()`.

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+    debounceInput: utils._.debounce(function(e) {
+      this.address = normalise(e.target.value);
+    }, 110),
+    copyToClipboard(ref) {
+      this.$refs[ref].select();
+      document.execCommand('copy');
+    },
+    async createTx() {
+      this.nonce = await this.$store.state.web3.eth.getTransactionCount(
+        this.wallet.getAddressString(),
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: replace: `web3.eth.getCoinbase()`

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+    },
+    setSelectedCurrency(e) {
+      this.selectedCurrency = e;
+    },
+    estimateGas() {
+      const isEth = this.selectedCurrency.name === 'Ethereum';
+      const bnAmount = new BigNumber(this.amount);
+      this.web3.eth
+        .estimateGas({
+          from: this.wallet.getAddressString(),
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: replace: `web3.eth.getCoinbase()`

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+    }, 501),
+    debounceInput: utils._.debounce(function(e) {
+      this.address = normalise(e.target.value);
+    }, 110),
+    copyToClipboard(ref) {
+      this.$refs[ref].select();
+      document.execCommand('copy');
+    },
+    async createTx() {
+      this.nonce = await this.$store.state.web3.eth.getTransactionCount(
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: since web3 has been imported through the getters, you can just shorten this by using `this.web3.eth`

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+        this.wallet.getAddressString(),
+        'latest'
+      );
+      const value = this.amount === '' ? 0 : unit.toWei(this.amount, 'ether');
+      const CoralSafeSendContract = new this.$store.state.web3.eth.Contract(
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: Use imported state variable: `this.web3.eth`

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+      const query = CoralSafeSendContract.methods['deposit'](
+        to,
+        protectionLevel
+      );
+      const encodedABI = query.encodeABI();
+      const gasLimit = CoralConfig.gasLimitSuggestion;
+      const valueLessGas =
+        parseFloat(value, 10) - Number(unit.toWei(gasLimit, 'gwei'));
+      this.raw = {
+        from: this.$store.state.wallet.getAddressString(),
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: replace with `this.web3.eth.getCoinbase()`

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 16]
```diff
@@ -0,0 +1,460 @@
...
+        value: valueLessGas,
+        to: CoralConfig.safeSendEscrowContractAddress,
+        nonce: this.nonce,
+        gas: gasLimit,
+        data: encodedABI,
+        gasPrice: Number(unit.toWei(this.$store.state.gasPrice, 'gwei')),
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: use `this.gasPrice` since it's imported already

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 17]
```diff
@@ -0,0 +1,460 @@
...
+        chainId: CoralConfig.chainID || 1
+      };
+
+      if (this.address === '') {
+        delete this.raw['to'];
+      }
+      this.$store.state.web3.eth.sendTransaction(this.raw);
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: `this.web3.eth`

File: src/dapps/SecureTransaction/SecureTransaction.vue [hunk 18]
```diff
@@ -0,0 +1,460 @@
...
+      }
+      this.$store.dispatch('setGasPrice', Number(newVal));
+    },
+    selectedCurrency(newVal) {
+      this.selectedCurrency = newVal;
+      this.estimateGas();
+    }
+  },
+  mounted() {
+    const address = this.$store.state.wallet.getAddressString();
```
Comment (gamalielhere) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: `this.web3.eth.getCoinbase()`

File: src/dapps/SecureTransaction/SecureTransaction.vue (hunk 10 above)
Comment (kvhnuke) [inline src/dapps/SecureTransaction/SecureTransaction.vue]: can you replace the instances of `wallet` usage to `await web3.eth.getCoinbase()` also take a look at our new send Transaction codebase, which is lot cleaner https://github.com/MyEtherWallet/MyEtherWallet/blob/develop/src/layouts/InterfaceLayout/containers/SendCurrencyContainer/SendCurrencyContainer.vue

File: src/dapps/SecureTransaction/SecureTransaction.vue (hunk 16 above)
Comment (kvhnuke) [reply src/dapps/SecureTransaction/SecureTransaction.vue]: also do `new BigNumber(unit.toWei(this.gasPrice, 'gwei')).toString()`

**Comment by austinbourdier** [comment]
@kvhnuke @gamalielhere updated!

**Comment by gamalielhere** [comment]
closing this for #594
